### PR TITLE
Adds custom executor with config management

### PR DESCRIPTION
**Pull request by NimaVaziri**
This enables the custom executor to run by default and provides the configuration management feature on the executor side (as needed to finish off the Task Config routing #198).

The new hierarchy of the project now follows this pattern:
```
sdk/
  |-core
  |-executor
frameworks/
  |-elasticsearch/
  |-presto
examples/
  |-reference
docs/
```

The top-level gradle build will build the subprojects, all of which will depend on the same version of the sdk (as referred to by `core`). For now just ported the reference example, but more will follow up soon.

A subproject can be built independently via `./gradlew build -p <subproject-dir-path>`

Combines #221 and #212.

### CODE REVIEW

**Comment by orsenthil** [comment]
```The top-level gradle build will build the subprojects, all of which will depend on the same version of the sdk (as referred to by core). For now just ported the reference example, but more will follow up soon.```

Chiming in from a build system perspective.

How does the top-level gradle build perform with respect to speed? Is the cost-incurred tolerable?

**Comment by NimaVaziri** [comment]
As of right now, the 4 projects collectively take 1m7s for a clean build, albeit `sdk/core` is the only project with 80%+ test coverage (which takes ~49s for a clean build).

File: sdk/executor/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#Fri Oct 21 19:20:13 PDT 2016
```
Comment (nickbp) [inline sdk/executor/gradle/wrapper/gradle-wrapper.properties]: Does this directory need its own copy of the gradle bootstrap stuff?

**Comment by nickbp** [comment]
LG overall, just the one comment about extra gradle bootstrap files and the needed rebase

File: sdk/executor/gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (NimaVaziri) [inline sdk/executor/gradle/wrapper/gradle-wrapper.properties]: Nope, fixed.

File: examples/reference/README.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
+Building a Stateful DC/OS Service
```
Comment (gabrielhartmann) [inline examples/reference/README.md]: I would delete this whole file, or replace it with a single reference to the new tutorial doc.  This is out of date.

File: examples/reference/build.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+# Grab dcos-commons build/release tools:
+rm -rf dcos-commons-tools/ && curl https://infinity-artifacts.s3.amazonaws.com/dcos-commons-tools.tgz | tar xz
+
+# GitHub notifier config
+_notify_github() {
+    $REPO_ROOT_DIR/dcos-commons-tools/github_update.py $1 build $2
+}
+
+# Build steps for Kafka
```
Comment (gabrielhartmann) [inline examples/reference/build.sh]: Shouldn't reference Kafka anymore

File: examples/reference/src/main/java/com/mesosphere/sdk/reference/scheduler/Main.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+            @Override
+            public List<TaskSet> getTaskSets() {
+                return Arrays.asList(
+                        DefaultTaskSet.create(TASK_METADATA_COUNT,
+                                TASK_METADATA_NAME,
+                                getCommand(TASK_METADATA_NAME),
+                                getResources(TASK_METADATA_CPU, TASK_METADATA_MEM_MB),
+                                getVolumes(TASK_METADATA_DISK_MB, TASK_METADATA_NAME),
+                                new ArrayList<>(),
```
Comment (gabrielhartmann) [inline examples/reference/src/main/java/com/mesosphere/sdk/reference/scheduler/Main.java]: `Collestions.emptyList()`

File: examples/reference/src/main/java/com/mesosphere/sdk/reference/scheduler/Main.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                                // avoid colocating with other metadata instances (<=1 instance/agent):
+                                Optional.of(TaskTypeGenerator.createAvoid(TASK_METADATA_NAME)),
+                                Optional.empty()),
+                        DefaultTaskSet.create(TASK_DATA_COUNT,
+                                TASK_DATA_NAME,
+                                getCommand(TASK_DATA_NAME),
+                                getResources(TASK_DATA_CPU, TASK_DATA_MEM_MB),
+                                getVolumes(TASK_DATA_DISK_MB, TASK_DATA_NAME),
+                                new ArrayList<>(),
```
Comment (gabrielhartmann) [inline examples/reference/src/main/java/com/mesosphere/sdk/reference/scheduler/Main.java]: `Collestions.emptyList()`

File: sdk/core/src/main/java/org/apache/mesos/offer/DefaultOfferRequirementProvider.java [hunk 6]
```diff
@@ -187,4 +189,35 @@ private static void validateVolumes(Protos.TaskInfo taskInfo, TaskSpecification
...
+        String executorStr = System.getenv(EXECUTOR_URI);
+        if (executorStr == null) {
+            throw new IllegalStateException("Missing environment variable: " + EXECUTOR_URI);
+        }
+        executorURI = TaskUtils.uri(executorStr);
+
+        Protos.CommandInfo.Builder commandInfoBuilder = Protos.CommandInfo.newBuilder()
+                .setValue("./executor/bin/executor")
+                .addUris(executorURI)
+                .setEnvironment(taskSpecification.getCommand().getEnvironment())
```
Comment (gabrielhartmann) [inline sdk/core/src/main/java/org/apache/mesos/offer/DefaultOfferRequirementProvider.java]: Was this resolved?

Comment (NimaVaziri) [inline sdk/core/src/main/java/org/apache/mesos/offer/DefaultOfferRequirementProvider.java]: Yup, don't need it. Removed.

**Pull request merged**